src/main/java/uk/gov/ida/verifyserviceprovider/metadata/MetadataPublicKeyExtractor.java

```diff
@@ -0,0 +1,72 @@
+package uk.gov.ida.verifyserviceprovider.metadata;
+
+import net.shibboleth.utilities.java.support.resolver.CriteriaSet;
+import net.shibboleth.utilities.java.support.resolver.ResolverException;
+import org.opensaml.core.criterion.EntityIdCriterion;
+import org.opensaml.saml.common.xml.SAMLConstants;
+import org.opensaml.saml.metadata.resolver.MetadataResolver;
+import org.opensaml.saml.saml2.metadata.EntityDescriptor;
+import org.opensaml.saml.saml2.metadata.KeyDescriptor;
+import org.opensaml.security.credential.UsageType;
+import org.opensaml.xmlsec.signature.X509Certificate;
+import uk.gov.ida.saml.security.PublicKeyFactory;
+
+import java.security.PublicKey;
+import java.text.MessageFormat;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+public class MetadataPublicKeyExtractor {
+
+    private final String entityId;
+    private final MetadataResolver metadataResolver;
+    private PublicKeyFactory publicKeyFactory;
+
+    public MetadataPublicKeyExtractor(
+        String entityId,
+        MetadataResolver metadataResolver,
+        PublicKeyFactory publicKeyFactory
+    ) {
+        this.entityId = entityId;
+        this.metadataResolver = metadataResolver;
+        this.publicKeyFactory = publicKeyFactory;
+    }
+
+    public PublicKey getEncryptionPublicKey() {
+        try {
+            CriteriaSet criteria = new CriteriaSet(new EntityIdCriterion(entityId));
+            return Optional.ofNullable(metadataResolver.resolveSingle(criteria))
+                .map(this::getPublicKeys)
+                .orElseThrow(missingEntityIdException());
```

Comment:
I think this would be slightly neater as `.orElseThrow(this::missingEntityIdException)` where `missingEntityIdException` produces a `RuntimeException` instead of a `Supplier<RuntimeException>`.